components/resolved_ts/src/advance.rs

```diff
@@ -227,9 +230,10 @@ impl<T: 'static + RaftStoreRouter<E>, E: KvEngine> AdvanceTsWorker<T, E> {
                                 continue;
                             }
                             let mut read_state = ReadState::default();
-                            if let Some(rp) = meta.region_read_progress.get(&region_id) {
-                                read_state.set_applied_index(rp.applied_index());
-                                read_state.set_safe_ts(rp.safe_ts());
+                            if let Some(rrp) = meta.region_read_progress.get(&region_id) {
+                                let rs = rrp.read_state();
```

Comment:
The observe region is only registered for the leader, so the lock will be accessed `leader count` times. Although the `leader count` could be large as the `region count` grow, I think it is okay since it is executed in a dedicated thread with a default interval of 1s.

Reply:
Then it's `peer count * leader count`. Why not get the read state before looping at L224?

Reply:
Done